src/aten/sycl/SummaryOpsKernel.cpp

```diff
@@ -0,0 +1,300 @@
+#pragma GCC diagnostic ignored "-Wreturn-type"
+
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+#include <aten/sycl/Atomics.h>
+#include <comm/Runtime.h>
+#include <comm/SYCLContext.h>
+#include <comm/SYCLHelpers.h>
+#include <comm/TensorInfo.h>
+
+namespace at::native::xpu {
+using namespace at::native;
+using namespace at::xpu::detail;
+
+template <typename input_t, typename IndexType>
+static IndexType getBin(
+    input_t bVal,
+    at::acc_type<input_t, true> minvalue,
+    at::acc_type<input_t, true> maxvalue,
+    int nbins) {
+  IndexType bin = (int)((bVal - minvalue) * nbins / (maxvalue - minvalue));
+  // (only applicable for histc)
+  // while each bin is inclusive at the lower end and exclusive at the higher,
+  // i.e. [start, end)
+  // the last bin is inclusive at both, i.e. [start, end], in order to include
+  // maxvalue if exists
+  // therefore when bin == nbins, adjust bin to the last bin
+  if (bin == nbins){
+    bin -= 1;
+  }
+    
+  return bin;
+}
+
+template <
+    typename output_t,
+    typename input_t,
+    typename IndexType,
+    int ADims,
+    bool has_weight,
+    typename Op>
+struct Histogram1DKernelFunctor {
+  void operator()(sycl::item<1> item_id) const {
+    auto out_ptr = out_data;
+    auto in_ptr = in_data;
+    auto weight_ptr = weight_data;
+
+    auto linearIndex = item_id.get_id(0);
+    // Convert `linearIndex` into an offset of `b`
+    const IndexType bOffset =
+        IndexToOffset<input_t, IndexType>::get(linearIndex, b);
+    const auto bVal = in_ptr[bOffset];
+    if (bVal >= minvalue && bVal <= maxvalue) {
+      // Use value at `b` as an offset of `a`
+      const IndexType bin =
+          getBin<input_t, IndexType>(bVal, minvalue, maxvalue, nbins);
+      const IndexType aOffset = IndexToOffset<output_t, IndexType>::get(bin, a);
+      atomicAdd(
+          (sycl_global_ptr<output_t>) &out_ptr[aOffset],
+          getOp(weight_ptr, linearIndex));
+    }
+  }
+  Histogram1DKernelFunctor(
+      TensorInfo<output_t, IndexType> a_,
+      TensorInfo<input_t, IndexType> b_,
+      TensorInfo<output_t, IndexType> c_,
+      int nbins_,
+      at::acc_type<input_t, true> minvalue_,
+      at::acc_type<input_t, true> maxvalue_,
+      IndexType totalElements_,
+      Op getOp_,
+      output_t* out_data_,
+      input_t* in_data_,
+      output_t* weight_data_)
+      : a(a_),
+        b(b_),
+        c(c_),
+        nbins(nbins_),
+        minvalue(minvalue_),
+        maxvalue(maxvalue_),
+        totalElements(totalElements_),
+        getOp(getOp_),
+        out_data(out_data_),
+        in_data(in_data_),
+        weight_data(weight_data_) {}
+
+ private:
+  TensorInfo<output_t, IndexType> a;
+  TensorInfo<input_t, IndexType> b;
+  TensorInfo<output_t, IndexType> c;
+  int nbins;
+  at::acc_type<input_t, true> minvalue;
+  at::acc_type<input_t, true> maxvalue;
+  IndexType totalElements;
+  Op getOp;
+  output_t* out_data;
+  input_t* in_data;
+  output_t* weight_data;
```

Comment:
Why these three? Data ptrs should be included in a, b, c?

Reply:
Regarding optimization guideline, we always tend to reduce arguments to offload to mitigate register pressure.